Cascade Delete Results In NullPointerException

When issuing a Cascade Delete, a NullPointerException is encountered when attempts are made to delete referenced Resources.

The following Stack Trace has been modified to only provide the relevant Open Source references:
```
13:30:13.634 [fhir_endpoint.fhirEndpointServer-2208] ERROR c.u.f.r.s.i.ExceptionHandlingInterceptor - Failure during REST processing
ca.uhn.fhir.rest.server.exceptions.InternalErrorException: Failed to call access method: java.lang.NullPointerException
	:   :   : 
ca.uhn.fhir.rest.server.method.BaseOutcomeReturningMethodBinding.invokeServer(BaseOutcomeReturningMethodBinding.java:154)
	at ca.uhn.fhir.rest.server.method.BaseOutcomeReturningMethodBindingWithResourceIdButNoResourceBody.invokeServer(BaseOutcomeReturningMethodBindingWithResourceIdButNoResourceBody.java:33)
	at ca.uhn.fhir.rest.server.RestfulServer.handleRequest(RestfulServer.java:1062)
	at ca.uhn.fhir.rest.server.RestfulServer.doDelete(RestfulServer.java:366)
	:   :   :
Caused by: java.lang.reflect.InvocationTargetException: null
	at ca.uhn.fhir.rest.server.method.BaseMethodBinding.invokeServerMethod(BaseMethodBinding.java:245) [1 skipped]
	at ca.uhn.fhir.rest.server.method.BaseOutcomeReturningMethodBinding.invokeServer(BaseOutcomeReturningMethodBinding.java:154)
	... 106 common frames omitted
Caused by: java.lang.NullPointerException: null
	at ca.uhn.fhir.rest.server.interceptor.partition.RequestTenantPartitionInterceptor.extractPartitionIdFromRequest(RequestTenantPartitionInterceptor.java:64)
	at ca.uhn.fhir.rest.server.interceptor.partition.RequestTenantPartitionInterceptor.PartitionIdentifyRead(RequestTenantPartitionInterceptor.java:57)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at ca.uhn.fhir.interceptor.executor.BaseInterceptorService$HookInvoker.invoke(BaseInterceptorService.java:530) [1 skipped]
	at ca.uhn.fhir.interceptor.executor.BaseInterceptorService.doCallHooks(BaseInterceptorService.java:291)
	at ca.uhn.fhir.interceptor.executor.BaseInterceptorService.callHooksAndReturnObject(BaseInterceptorService.java:260)
	at ca.uhn.fhir.jpa.util.JpaInterceptorBroadcaster.doCallHooksAndReturnObject(JpaInterceptorBroadcaster.java:62)
	at ca.uhn.fhir.jpa.partition.RequestPartitionHelperSvc.determineReadPartitionForRequest(RequestPartitionHelperSvc.java:109)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.readEntity(BaseHapiFhirResourceDao.java:1145)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.readEntity(BaseHapiFhirResourceDao.java:1131)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.doRead(BaseHapiFhirResourceDao.java:1088)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.lambda$read$9(BaseHapiFhirResourceDao.java:1074)
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.execute(HapiTransactionService.java:65) [1 skipped]
	at ca.uhn.fhir.jpa.dao.tx.HapiTransactionService.execute(HapiTransactionService.java:56)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.read(BaseHapiFhirResourceDao.java:1074)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.read(BaseHapiFhirResourceDao.java:1067)
	at ca.uhn.fhir.jpa.dao.BaseHapiFhirResourceDao.read(BaseHapiFhirResourceDao.java:1062)
	... 111 common frames omitted
```

--- a/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptor.java
+++ b/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptor.java
@@ -129,7 +129,7 @@ public class CascadingDeleteInterceptor {
 				IFhirResourceDao dao = myDaoRegistry.getResourceDao(nextSource.getResourceType());
 
 				// Interceptor call: STORAGE_CASCADE_DELETE
-				IBaseResource resource = dao.read(nextSource);
+				IBaseResource resource = dao.read(nextSource, theRequest);
 				HookParams params = new HookParams()
 					.add(RequestDetails.class, theRequest)
 					.addIfMatchesType(ServletRequestDetails.class, theRequest)

--- a/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptorTest.java
+++ b/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptorTest.java
@@ -3,9 +3,7 @@ package ca.uhn.fhir.jpa.interceptor;
 import ca.uhn.fhir.interceptor.api.IInterceptorBroadcaster;
 import ca.uhn.fhir.jpa.api.dao.DaoRegistry;
 import ca.uhn.fhir.jpa.api.dao.IFhirResourceDao;
-import ca.uhn.fhir.jpa.dao.BaseHapiFhirDao;
 import ca.uhn.fhir.jpa.provider.r4.BaseResourceProviderR4Test;
-import ca.uhn.fhir.model.primitive.IdDt;
 import ca.uhn.fhir.rest.api.Constants;
 import ca.uhn.fhir.rest.api.server.RequestDetails;
 import ca.uhn.fhir.rest.server.exceptions.ResourceGoneException;
@@ -28,16 +26,19 @@ import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Test;
 import org.mockito.ArgumentCaptor;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.testcontainers.shaded.org.bouncycastle.cert.ocsp.Req;
 
 import java.io.IOException;
 import java.util.List;
 
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.fail;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.*;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
 public class CascadingDeleteInterceptorTest extends BaseResourceProviderR4Test {
 

--- a/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptorTest.java
+++ b/hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/CascadingDeleteInterceptorTest.java
@@ -2,14 +2,19 @@ package ca.uhn.fhir.jpa.interceptor;
 
 import ca.uhn.fhir.interceptor.api.IInterceptorBroadcaster;
 import ca.uhn.fhir.jpa.api.dao.DaoRegistry;
+import ca.uhn.fhir.jpa.api.dao.IFhirResourceDao;
+import ca.uhn.fhir.jpa.dao.BaseHapiFhirDao;
 import ca.uhn.fhir.jpa.provider.r4.BaseResourceProviderR4Test;
+import ca.uhn.fhir.model.primitive.IdDt;
 import ca.uhn.fhir.rest.api.Constants;
+import ca.uhn.fhir.rest.api.server.RequestDetails;
 import ca.uhn.fhir.rest.server.exceptions.ResourceGoneException;
 import ca.uhn.fhir.rest.server.exceptions.ResourceVersionConflictException;
 import com.google.common.base.Charsets;
 import org.apache.commons.io.IOUtils;
 import org.apache.http.client.methods.CloseableHttpResponse;
 import org.apache.http.client.methods.HttpDelete;
+import org.hl7.fhir.instance.model.api.IBaseResource;
 import org.hl7.fhir.instance.model.api.IIdType;
 import org.hl7.fhir.r4.model.CarePlan;
 import org.hl7.fhir.r4.model.Condition;
@@ -21,14 +26,18 @@ import org.hl7.fhir.r4.model.Patient;
 import org.hl7.fhir.r4.model.Reference;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.testcontainers.shaded.org.bouncycastle.cert.ocsp.Req;
 
 import java.io.IOException;
+import java.util.List;
 
 import static org.hamcrest.CoreMatchers.containsString;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.fail;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
 
 public class CascadingDeleteInterceptorTest extends BaseResourceProviderR4Test {
 
@@ -86,6 +95,42 @@ public class CascadingDeleteInterceptorTest extends BaseResourceProviderR4Test {
 		condition.setAsserter(new Reference(myPatientId));
 		condition.setEncounter(new Reference(myEncounterId));
 		myConditionId = myClient.create().resource(condition).execute().getId().toUnqualifiedVersionless();
+	}
+
+	@Test
+	public void testDeleteWithInterceptorVerifyTheRequestGetsPassedToDao() throws IOException {
+		// The whole and ONLY point of this Cascade Delete Unit Test is to make sure that a non-NULL RequestDetails param
+		// is passed to the dao.read() method from inside the CascadingDeleteInterceptor.handleDeleteConflicts() method
+		// For details see: https://gitlab.com/simpatico.ai/cdr/-/issues/1643
+		DaoRegistry mockDaoRegistry = mock(DaoRegistry.class);
+		IFhirResourceDao mockResourceDao = mock (IFhirResourceDao.class);
+		IBaseResource mockResource = mock(IBaseResource.class);
+		CascadingDeleteInterceptor aDeleteInterceptor = new CascadingDeleteInterceptor(myFhirCtx, mockDaoRegistry, myInterceptorBroadcaster);
+		ourRestServer.getInterceptorService().unregisterInterceptor(myDeleteInterceptor);
+		ourRestServer.getInterceptorService().registerInterceptor(aDeleteInterceptor);
+		when(mockDaoRegistry.getResourceDao(any(String.class))).thenReturn(mockResourceDao);
+		when(mockResourceDao.read(any(IIdType.class), any(RequestDetails.class))).thenReturn(mockResource);
+		ArgumentCaptor<RequestDetails> theRequestDetailsCaptor = ArgumentCaptor.forClass(RequestDetails.class);
+
+		Patient p = new Patient();
+		p.setActive(true);
+		myPatientId = myClient.create().resource(p).execute().getId().toUnqualifiedVersionless();
+		Encounter e = new Encounter();
+		e.setSubject(new Reference(myPatientId));
+		myEncounterId = myClient.create().resource(e).execute().getId().toUnqualifiedVersionless();
+
+		HttpDelete delete = new HttpDelete(ourServerBase + "/" + myPatientId.getValue() + "?" + Constants.PARAMETER_CASCADE_DELETE + "=" + Constants.CASCADE_DELETE + "&_pretty=true");
+		delete.addHeader(Constants.HEADER_ACCEPT, Constants.CT_FHIR_JSON_NEW);
+		try (CloseableHttpResponse response = ourHttpClient.execute(delete)) {
+			String deleteResponse = IOUtils.toString(response.getEntity().getContent(), Charsets.UTF_8);
+			ourLog.info("Response: {}", deleteResponse);
+		}
+
+		verify(mockResourceDao).read(any(IIdType.class), theRequestDetailsCaptor.capture());
+		List<RequestDetails> capturedRequestDetailsParam = theRequestDetailsCaptor.getAllValues();
+		for (RequestDetails requestDetails : capturedRequestDetailsParam) {
+			assertNotNull(requestDetails);
+		}
 	}
 
 	@Test
@@ -252,7 +297,4 @@ public class CascadingDeleteInterceptorTest extends BaseResourceProviderR4Test {
 			// good
 		}
 	}
-
-
-
 }

